jax/scipy/sparse/linalg.py

```diff
@@ -54,42 +61,123 @@ def _identity(x):
 def _cg_solve(A, b, x0=None, *, maxiter, tol=1e-5, atol=0.0, M=_identity):
 
   # tolerance handling uses the "non-legacy" behavior of scipy.sparse.linalg.cg
-  bs = _vdot_tree(b, b)
+  bs = _vdot_tree_real(b, b)
   atol2 = jnp.maximum(jnp.square(tol) * bs, jnp.square(atol))
 
   # https://en.wikipedia.org/wiki/Conjugate_gradient_method#The_preconditioned_conjugate_gradient_method
 
   def cond_fun(value):
     x, r, gamma, p, k = value
-    rs = gamma if M is _identity else _vdot_tree(r, r)
+    rs = gamma if M is _identity else _vdot_tree_real(r, r)
     return (rs > atol2) & (k < maxiter)
 
   def body_fun(value):
     x, r, gamma, p, k = value
     Ap = A(p)
-    alpha = gamma / _vdot_tree(p, Ap)
+    alpha = gamma / _vdot_tree_real(p, Ap)
     x_ = _add(x, _mul(alpha, p))
     r_ = _sub(r, _mul(alpha, Ap))
     z_ = M(r_)
-    gamma_ = _vdot_tree(r_, z_)
+    gamma_ = _vdot_tree_real(r_, z_)
     beta_ = gamma_ / gamma
     p_ = _add(z_, _mul(beta_, p))
     return x_, r_, gamma_, p_, k + 1
 
   r0 = _sub(b, A(x0))
   p0 = z0 = M(r0)
-  gamma0 = _vdot_tree(r0, z0)
+  gamma0 = _vdot_tree_real(r0, z0)
   initial_value = (x0, r0, gamma0, p0, 0)
 
   x_final, *_ = lax.while_loop(cond_fun, body_fun, initial_value)
 
   return x_final
 
 
+def _bicgstab_solve(A, b, x0=None, *, maxiter, tol=1e-5, atol=0.0, M=_identity):
+
+  # tolerance handling uses the "non-legacy" behavior of scipy.sparse.linalg.bicgstab
+  bs = _vdot_tree_real(b, b)
+  atol2 = jnp.maximum(jnp.square(tol) * bs, jnp.square(atol))
+
+  # https://en.wikipedia.org/wiki/Biconjugate_gradient_stabilized_method#Preconditioned_BiCGSTAB
+
+  def cond_fun(value):
+    x, r, *_, k = value
+    rs = _vdot_tree_real(r, r)
+    return (rs > atol2) & (k < maxiter)
+
+  def body_fun(value):
+    x, r, rhat, alpha, omega, rho, p, q, k = value
+    # rho_ = _vdot_tree(_conj(rhat), r)
+    rho_ = _vdot_tree(rhat, r)
+    beta = rho_ / rho * alpha / omega
+    p_ = _add(r, _mul(beta, _sub(p, _mul(omega, q))))
+    phat = M(p_)
+    q_ = A(phat)
+    # alpha_ = rho_ / _vdot_tree(_conj(rhat), q_)
+    alpha_ = rho_ / _vdot_tree(rhat, q_)
+    s = _sub(r, _mul(alpha_, q_))
+    # TODO(?): stop early
+    #  It requires accessing cond_fun like this, but it's
+    #  not possible with jit...
+    #  if cond_fun((x, s, r0, alpha_, omega, rho_, p_, q_, k)):
+    #    x_ = _add(x, _mul(alpha_, phat))
+    #    return x_, s, rhat, alpha_, omega, rho_, p_, q, k
```

Comment:
We could probably do this with some cleverness (e.g., involving `lax.cond`), but I think it's actually totally fine to skip it -- at least for now. In the best case, this would save one extra function evaluation.

On GPUs, dynamic control flow is so expensive that I expect we could actually get better performance in many cases by only checking for stopping every `n` steps.